src/data/polynomial/hasse_deriv.lean

```diff
@@ -0,0 +1,220 @@
+/-
+Copyright (c) 2021 Johan Commelin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johan Commelin
+-/
+
+import data.polynomial.derivative
+import data.nat.choose.dvd
+import data.nat.choose.vandermonde
+import logic.function.iterate
+import tactic.ring
+
+/-!
+# Hasse derivative of polynomials
+
+The `k`th Hasse derivative of a polynomial `∑ a_i X^i` is `∑ (i.choose k) a_i X^(i-k)`.
+It is a variant of the usual derivative, and satisfies `k! * (hasse_deriv k f) = derivative^[k] f`.
+The main benefit is that is gives an atomic way of talking about expressions such as
+`(derivative^[k] f).eval r / k!`, that occur in Taylor expansions, for example.
+
+## Main declarations
+
+In the following, we write `D k` for the `k`-th Hasse derivative `hasse_deriv k`.
+
+* `polynomial.hasse_deriv`: the `k`-th Hasse derivative of a polynomial
+* `polynomial.hasse_deriv_zero`: the `0`th Hasse derivative is the identity
+* `polynomial.hasse_deriv_one`: the `1`st Hasse derivative is the usual derivative
+* `polynomial.factorial_smul_hasse_deriv`: the identity `k! • (D k f) = derivative^[k] f`
+* `polynomial.hasse_deriv_comp`: the identity `(D k).comp (D l) = (k+l).choose k • D (k+l)`
+* `polynomial.hasse_deriv_mul`:
+  the "Leibniz rule" `D k (f * g) = ∑ ij in antidiagonal k, D ij.1 f * D ij.2 g`
+
+TODO: Prove the identity principle: a polynomial is 0 iff all its Hasse derivatives are zero.
```

Comment:
More importantly, prove the Taylor formula.

Reply:
See #9000 .